### ISSUE TITLE
ClientSidePage: Set LayoutWebPartsContent property for updating site banner

### ISSUE THREAD

**Issue by aakashbhardwaj619** [pnp/pnpjs#547]
### Category
- [x] Enhancement
- [ ] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: [1.3.0]

Please specify what version(s) of SharePoint you are targeting: [SharePoint Online]

### Expected / Desired Behavior / Question
I'm creating a copy of an existing ClientSidePage using the new copyPage method and its working fine. But I need to update the LayoutWebPartsContent property of this new page for updating the site banner image and text above title. In the previous version on pnp (1.2.7) I was doing this using the page.getItem method. Since this method is private now, I'm unable to use it in the new version to get and update the page item properties. Is there any alternative to do that?

`const page = await ClientSidePage.fromFile(sp.web.getFileByServerRelativePath("oldPageUrl"));
let newPage = await page.copyPage(sp.web, "copy.aspx", "Copied Title");

//These update the page properties but not the banner image and text
newPage.bannerImageUrl = "newImageUrl";
newPage.bannerImageSourceType = 2;
newPage.layoutType = "FullWidthImage";
newPage.topicHeader = "High Risk Work";
newPage.showTopicHeader = true;

newPage.save(true)`

**Comment by patrick-rodgers** [comment]
We'll have a look again at the properties - but as a quick work around you can always case the object to any and call the method as it is still there. At run time there is no type checking anyway.

**Comment by reshampal2010** [comment]
I am also facing similar problem, i have checked the value in "BannerImageUrl" from rest it's getting updated but image is not getting rendered.

**Comment by aakashbhardwaj619** [comment]
I just checked that I'm able to update the banner properties using spHttpClient and REST API **/_api/sitepages/page** endpoint (which is also being used in  ClientSidePage's property setters like topicHeader, bannerImageUrl, etc.). Not sure why the update stopped working if updating through list item's properties. Anyway I think if the **_layoutPart** member is made public then this can be easily achieved to update the LayoutWebPartsContent property.

```
//Banner property format can be get using any page's maintenance mode
let layoutJson = `[{ 
      "id": "cbe7b0a9-3504-44dd-a3a3-0e5cacd07788", 
	.
	. 
	.
    }]`;

//Updating topic header, banner image URL and LayoutWebPartsContent property
const body: string = JSON.stringify({
        '__metadata': {
          'type': 'SP.Publishing.SitePage'
        },
        'TopicHeader': subjectHeaderTopic,
        'BannerImageUrl': newPageBannerImage,
        'LayoutWebpartsContent': layoutJson
      });

      /**
       * Checkout the new page and update banner properties using REST API
       */
      props.context.spHttpClient.post(`${props.context.pageContext.web.absoluteUrl}/_api/sitepages/pages(${newPageItemId})/checkoutpage`, SPHttpClient.configurations.v1, {}).then((data: SPHttpClientResponse) => {
        console.log('Page checked out');
        if (data.ok) {
          props.context.spHttpClient.post(`${props.context.pageContext.web.absoluteUrl}/_api/sitepages/pages(${newPageItemId})`,
            SPHttpClient.configurations.v1,
            {
              headers: {
                'Accept': 'application/json;odata=nometadata',
                'Content-type': 'application/json;odata=verbose',
                'odata-version': '',
                'IF-MATCH': '*',
                'X-HTTP-Method': 'MERGE'
              },
              body: body
            }).then(() => {
              console.log("Page Banner updated");
            });
        }
      });
```

**Comment by patrick-rodgers** [comment]
Hi @aakashbhardwaj619 - circling back on this, the properties seem to work for me to set things such as topicHeader etc. Do you know what was different in the request we are sending and the one you sent? Are you setting all the extra metadata? That is something we aren't currently doing.

**Comment by patrick-rodgers** [comment]
Did find a bug in setting banner image, we missed setting the value also in server processed content. So I think that is fixed now. All the other properties seem to be working as intended. This fix will be included in the next beta. I am also going to add a set banner image method to expose all those various values instead of using the properties.

**Comment by patrick-rodgers** [comment]
Closing this as fixed, will be in the next beta as well as release. Thanks for reporting the issue!

**Comment by patrick-rodgers** [comment]
Just published a new beta, v1.3.1-1, which contains the updates mentioned in this issue. Please give it a try if you have time and let us know if you see any further issues. Thanks!

**Comment by aakashbhardwaj619** [comment]
I just tested the new **setBannerImage** method. The method works fine for replacing the banner image of a page that already a different image set as the banner image. However, for a newly created page or a page that does not have an initial banner image set the method gives the error message: **TypeError: Cannot convert undefined or null to object**

I think this might be because we are deleting the below properties in **bannerImageUrl** method that are undefined initially before a banner image is set:

```
delete this._layoutPart.properties.webId;
delete this._layoutPart.properties.siteId;
delete this._layoutPart.properties.listId;
delete this._layoutPart.properties.uniqueId;
```
